polaris/competition/_base.py

```diff
@@ -0,0 +1,229 @@
+        input_cols: The column(s) of the original dataset that should be used as input.
+        split: The predefined train-test split to use for evaluation.
+        metrics: The metrics to use for evaluating performance
+        main_metric: The main metric used to rank methods.
+        readme: Markdown text that contains a formatted description of the competition.
+        target_types: A dictionary that maps target columns to their type.
+        start_time: The time at which the competition starts accepting prediction submissions.
+        end_time: The time at which the competition stops accepting prediction submissions.
+        n_test_sets: The number of test sets defined in the test split.
+        n_test_datapoints: The number of datapoints in each test set.
+        n_classes: The number of classes within target columns that define a classification task.
+    For additional meta-data attributes, see the [`DatasetV2`][polaris.experimental._dataset_v2.DatasetV2] class.
+    """
+
+    _artifact_type = "competition"
+
+    target_cols: ColumnsType
+    input_cols: ColumnsType
+    split: SplitType
+    metrics: Union[str, Metric, list[str | Metric]]
+    main_metric: str | Metric
+    readme: str
+    target_types: dict[str, Union[TargetType, str]] = Field(default_factory=dict, validate_default=True)
+    start_time: datetime
+    end_time: datetime
+    n_test_sets: int
+    n_test_datapoints: dict[str, int]
+    n_classes: dict[str, int]
+
+    @field_validator("metrics")
+    def _validate_metrics(cls, v) -> list[Metric]:
+        """
+        Verifies all specified metrics are either a Metric object or a valid metric name.
+
+        If there are multiple test sets, it is assumed the same metrics are used across test sets.
+        """
+        if not isinstance(v, list):
+            v = [v]
+
+        v = [m if isinstance(m, Metric) else Metric[m] for m in v]
+
+        return v
+
+    @field_validator("main_metric")
+    def _validate_main_metric(cls, v) -> Metric:
+        """Converts the main metric to a Metric object if it is a string."""
+        # lu: v can be None.
+        if v and not isinstance(v, Metric):
+            v = Metric[v]
+        return v
+
+    @field_serializer("metrics", "main_metric")
+    def _serialize_metrics(self, v) -> list[str] | str:
+        """Return the string identifier so we can serialize the object"""
+        if isinstance(v, Metric):
+            return v.name
+        return [m.name for m in v]
+
+    @field_serializer("target_types")
+    def _serialize_target_types(self, v) -> dict[str, str]:
+        """Convert from enum to string to make sure it's serializable"""
+        return {k: v for k, v in self.target_types.items()}
+
+    @field_serializer("start_time", "end_time")
+    def _serialize_times(self, v: datetime) -> str:
+        """Convert from datetime to string to make sure it's serializable"""
+        return v.isoformat()
```

Comment:
**Suggestion:** I believe you can drop this. Pydantic got you covered! See e.g. [here](https://docs.pydantic.dev/2.2/usage/serialization/#modelmodel_dump_json)

```py
from datetime import datetime
from pydantic import BaseModel

class FooModel(BaseModel):
    foo: datetime

m = FooModel(foo=datetime(2032, 6, 1, 12, 13, 14))
print(m.model_dump_json())
#> {"foo":"2032-06-01T12:13:14"}
```

I do think we need to then switch to using `model_dump_json`, rather than `model_dump`, but I actually think our current use of `model_dump` is not correct!